File: barbican/tree/branch/display.py
from tree.branch.branch import Branch

from tree.file.display import display as display_file


def get_starter(index: int, count: int):
    if (index + 1) < count:
        return '├─ '
    else:
        return '╰─ '


def display(branch: Branch, leader: str = '', final: bool = True, root: bool = True):
    starter = '╰─ ' if final else '├─ '
    new_leader = '' if root else f'{leader}│  '

    if root:
        print(branch.name)
    else:
        print(leader, starter, branch.name, sep='')

    index = 0
    count = len(branch.headers) + len(branch.sources) + len(branch.branches)
    for branch_ in branch.branches.values():
        final = (index + 1) == count
        display(branch_, new_leader, final, root=False)
        index += 1

    for header in branch.headers.values():
        final = (index + 1) == count
        display_file(header, new_leader, final)
        index += 1

    for source in branch.sources.values():
        final = (index + 1) == count
        display_file(source, new_leader, final)
        index += 1

File: source/tree/branch/display.go
package branch

import (
	"fmt"

	"github.com/inigo-selwood/barbican/tree/asset/header"
	"github.com/inigo-selwood/barbican/tree/asset/source"
)

func Display(branch *Branch, leader string, final bool, root bool) {

	// Print starter
	starter := ""
	if root == false {
		if final {
			starter = "╰─ "
		} else {
			starter = "├─ "
		}
	}

	// Print the branch name
	fmt.Printf("%s%s%s\n", leader, starter, branch.Name)

	// Evaluate a new leader
	newLeader := ""
	if root == false {
		if final {
			newLeader = leader + "   "
		} else {
			newLeader = leader + "│  "
		}
	}

	// Print each child branch
	index := 0
	count := len(branch.Headers) + len(branch.Sources) + len(branch.Branches)
	for _, childBranch := range branch.Branches {
		finalBranch := (index + 1) == count
		Display(childBranch, newLeader, finalBranch, false)
		index += 1
	}

	// Print headers
	for _, instance := range branch.Headers {
		finalFile := (index + 1) == count
		header.Display(instance, newLeader, finalFile)
		index += 1
	}

	// Print sources
	for _, instance := range branch.Sources {
		finalFile := (index + 1) == count
		source.Display(instance, newLeader, finalFile)
		index += 1
	}
}
